### PR TITLE
Clean up sparse array API

### PR DESCRIPTION
**Pull request by stefanv**
Makes private many functions that are used internally. These remain exposed in the older matrix API.

### CODE REVIEW

File: scipy/sparse/_base.py [hunk 1]
```diff
@@ -63,6 +63,7 @@ class _sparray:
     """
 
     __array_priority__ = 10.1
+    _format = 'und'  # undefined
```
Comment (h-vetinari) [inline scipy/sparse/_base.py:66]: What do we gain by this abbreviation? Can we not just use
```suggestion
    _format = 'undefined'
```
? AFAICT this should only ever be in the base class and thus not visible to the users anyway.

Comment (stefanv) [reply scipy/sparse/_base.py:66]: I had that change originally, but then I saw that there is a dictionary of formats higher up in the file.

Comment (h-vetinari) [reply scipy/sparse/_base.py:66]: OK, thx.

**Comment by perimosocordiae** [comment]
One more test update: `test_array_api.py` is checking `getrow()` and `getcol()`, so we should either delete the test or replace the calls with the underscore versions.

Re: deprecation cycle, I'm hoping that we're still early enough in the lifecycle of the sparse array API that we can make some breaking changes without too much downstream breakage. In the 1.8.0 release notes (https://docs.scipy.org/doc/scipy/release.1.8.0.html#scipy-sparse-improvements), we wrote:

> An array API has been added for early testing and feedback; this work is ongoing, and users should expect minor API refinements over the next few releases. Please refer to the [scipy.sparse](https://docs.scipy.org/doc/scipy/reference/sparse.html#module-scipy.sparse) docstring for more information.

Now that we're looking at a 1.11 release these changes are a bit late, I admit.

**Comment by perimosocordiae** [comment]
After some in-person discussion, we decided to do proper deprecations for these.

File: scipy/sparse/_base.py [hunk 2]
```diff
@@ -1267,6 +1254,105 @@ def _process_toarray_args(self, order, out):
...
+        """See `reshape`."""
+        msg = "Shape assignment is deprecated. Use `reshape` instead."
+        warn(msg, DeprecationWarning, stacklevel=2)
+
+        # Make sure copy is False since this is in place
+        # Make sure format is unchanged because we are doing a __dict__ swap
+        new_self = self.reshape(shape, copy=False).asformat(self.format)
+        self.__dict__ = new_self.__dict__
+
+    shape = property(fget=lambda self: self._shape, fset=set_shape)  # noqa: F811
```
Comment (perimosocordiae) [inline scipy/sparse/_base.py:1277]: This is redefining `shape`, which causes mypy to complain in the CI

File: scipy/sparse/_base.py [hunk 3]
```diff
@@ -1267,6 +1254,105 @@ def _process_toarray_args(self, order, out):
             return np.zeros(self.shape, dtype=self.dtype, order=order)
 
 
+    ## All methods below are deprecated and should be released in
```
Comment (perimosocordiae) [inline scipy/sparse/_base.py:1257]: released -> removed, I assume?

**Comment by j-bowhay** [comment]
I would also prefer it if we added tests to make sure the deprecation warnings appear when expected using `pytest.deprecated_call():`

**Comment by stefanv** [comment]
OK done

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
Ralf's deprecation request has been handled, so I'll merge this now to unblock some upcoming PRs. If anyone has additional concerns, please let me know and I'll address them promptly.

**Comment by h-vetinari** [comment]
This PR has led to a docbuild failure:
```
[...]/scipy/sparse/_matrix.py:docstring of scipy.sparse._matrix.spmatrix.getnnz:16: WARNING: py:obj reference target not found: count_nonzero
```

**Comment by j-bowhay** [comment]
Thanks @stefanv for updating the messages and tests! I had one further comment to make but was waiting for the doc build to pass to confirm it. Currently, if I look at https://scipy.github.io/devdocs/reference/generated/scipy.sparse.bsr_array.get_shape.html#scipy.sparse.bsr_array.get_shape etc there is no indication to suggest this is deprecated so as a user I will get a bit of a surprise when I use it and get a deprecation warning. Would you be able to make a following up pr adding the `.. deprecated::` directive to the docstring of each of the deprecated functions with a similar message to what the warning emits?

**Comment by j-bowhay** [comment]
Also, it would be helpful if you could add an entry to the deprecation section of the release notes summarising what has been deprecated and what should be used instead

**Comment by h-vetinari** [comment]
> If anyone has additional concerns, please let me know and I'll address them promptly.

Could you please address the doc build failure soon? We've had a lot of issues with failing CI recently, it would be nice to have it back to green so shortly before release.

**Comment by rossbar** [comment]
> Could you please address the doc build failure soon? We've had a lot of issues with failing CI recently, it would be nice to have it back to green so shortly before release.

That should be fixed by #18526

**Pull request merged**